### PR TITLE
GH-91389: Fix `dis` position information for `CACHE`s

### PR DESCRIPTION
**Pull request by brandtbucher**
We currently don't advance the `co_positions` iterator for `CACHE` entries, which can lead to incorrect positions being reported for later instructions. Also, the positions for cache entries should be populated, not `None`.

Also, don't show a "pretty" `argrepr` for cache entries unless they're actually being used by a quickened instruction. This cuts down on visual clutter when `show_caches=True, adaptive=False`, or for instructionss we've completely unquickened.

### CODE REVIEW

File: Lib/test/test_dis.py [hunk 1]
```diff
@@ -1600,6 +1602,32 @@ def test_co_positions_missing_info(self):
...
+        code = roots.__code__
+        for show_caches in (False, True):
+            for adaptive in (False, True):
+                with self.subTest(f"{adaptive=}, {show_caches=}"):
+                    co_positions = [
+                        positions
+                        for op, positions in zip(
+                            code.co_code[::2], code.co_positions(), strict=True
+                        )
+                        if show_caches or op != opcode.opmap["CACHE"]
```
Comment (iritkatriel) [inline Lib/test/test_dis.py]: Minor: you could assign opcode.opmap["CACHE"] to a local outside the loop.

File: Lib/test/test_dis.py [hunk 2]
```diff
@@ -1600,6 +1602,32 @@ def test_co_positions_missing_info(self):
                 self.assertIsNone(positions.col_offset)
                 self.assertIsNone(positions.end_col_offset)
 
+    @requires_debug_ranges()
+    def test_co_positions_with_lots_of_caches(self):
+        def roots(a, b, c):
+            d = (b**2 - 4 * a * c) ** 0.5
+            yield (-b - d) / (2 * a)
+            if d:
+                yield (-b + d) / (2 * a)
```
Comment (iritkatriel) [inline Lib/test/test_dis.py]: Maybe add an assertion here that there are indeed "lots of caches"?

File: Lib/test/test_dis.py [hunk 3]
```diff
@@ -1751,6 +1753,36 @@ def test_co_positions_missing_info(self):
...
+    @requires_debug_ranges()
+    def test_co_positions_with_lots_of_caches(self):
+        def roots(a, b, c):
+            d = b**2 - 4 * a * c
+            yield (-b - cmath.sqrt(d)) / (2 * a)
+            if d:
+                yield (-b + cmath.sqrt(d)) / (2 * a)
+        code = roots.__code__
+        ops = code.co_code[::2]
+        cache = opcode.opmap["CACHE"]
```
Comment (iritkatriel) [inline Lib/test/test_dis.py]:
```suggestion
        cache_opcode = opcode.opmap["CACHE"]
```

File: Lib/test/test_dis.py [hunk 4]
```diff
@@ -1751,6 +1753,36 @@ def test_co_positions_missing_info(self):
...
+        non_caches = len(ops) - caches
+        # Make sure we have "lots of caches". If not, roots should be updated:
+        assert 1 / 3 <= caches / non_caches, "this test needs more caches!"
+        for show_caches in (False, True):
+            for adaptive in (False, True):
+                with self.subTest(f"{adaptive=}, {show_caches=}"):
+                    co_positions = [
+                        positions
+                        for op, positions in zip(ops, code.co_positions(), strict=True)
+                        if show_caches or op != cache
```
Comment (iritkatriel) [inline Lib/test/test_dis.py]:
```suggestion
                        if show_caches or op != cache_opcode
```

**Review by iritkatriel — APPROVED** [review]

**Comment by brandtbucher** [comment]
Closing and reopening to re-run CI.

**Pull request merged**